narciuch/assignment3/narciuch-3/src/WordSearch.java

```diff
@@ -0,0 +1,145 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class WordSearch {
+
+  /**
+   * Given a grid of letters and a dictionary, find all the words
+   * from the dictionary that can be formed in the grid.
+   *
+   * <p>If there are multiple instances of a word in the grid,
+   * then this word is returned multiple times in the result.
+   *
+   * <p>When forming a word, we assume:
+   * <ul>
+   * <li>Words can start at any position within the grid.</li>
+   * <li>We can move to one of the 8 adjacent cells (horizontally/vertically/diagonally)</li>
+   * <li>We cannot visit the same cell more than once in the same word.</li>
+   * </ul>
+   *
+   * @param grid of letters in which to look for words from the dictionary
+   * @param dict the dictionary of words to look for
+   * @return list of words from the dictionary that can be formed in the grid
+   */
+  public List<String> findWords(char[][] grid, Dictionary dict) {
+    List<String> wordsFound = new ArrayList<>();
+    if (grid.length == 0) {
```

Comment:
Good handling of a corner case!

Optional: handle the other corner case (`grid[0].length`) here explicitly as well rather than relying on loops not to process any elements. This would make code behaviour a bit more clear.

Reply:
Will do for clarity